src/prefect/engine/state.py

```diff
@@ -82,6 +84,40 @@ def __eq__(self, other: object) -> bool:
     def __hash__(self) -> int:
         return id(self)
 
+    def __getstate__(self) -> dict:
+        data = self.__dict__
+
+        # For 'Exception' results, we will check if it can be pickled successfully and
+        # on failure convert the exception to a string instead. This allows the engine
+        # to handle tasks with unpickable exceptions successfully. We must explicitly
+        # ignore state signals here or we will get stuck in a recursive loop because
+        # signals are exceptions that contain a state with `_result` set to the signal
+        if isinstance(self.result, Exception) and not isinstance(
+            self.result, prefect.engine.signals.PrefectStateSignal
+        ):
+
+            # Check for an exception during pickling of the 'Exception' type result
+            pickle_exc: Optional[Exception] = None
+            try:
+                cloudpickle.dumps(self.result)
+            except TypeError as exc:
+                if "cannot pickle" in str(exc):
+                    pickle_exc = exc
+            except PicklingError as exc:
+                pickle_exc = exc
+
```

Comment:
We may want to do...
```suggestion
            except PicklingError as exc:
                pickle_exc = exc
            except:
                pass
                # Any other exception will be ignored and we'll let the pickler fail
                # downstream
```

Reply:
Otherwise we're raising an exception early / if they're not using cloudpickle they could still succeed

Reply:
Nevermind... since the exception message changes across python versions and the exception types can vary we'll just catch any exception here as the pickle exception in https://github.com/PrefectHQ/prefect/pull/4605/commits/e44997652fbbe580a661b559c299284913b46dd6